### PR TITLE
Emit root vertex bft update on VertexStore rebuild

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt/src/main/java/com/radixdlt/consensus/bft/VertexStore.java [hunk 1]
```diff
@@ -66,36 +65,52 @@
...
 		VertexStoreEventSender vertexStoreEventSender,
 		SystemCounters counters
 	) {
-		this(rootVertex, rootQC, Collections.emptyList(), ledger, bftUpdateSender, vertexStoreEventSender, counters);
+		this.ledger = Objects.requireNonNull(ledger);
+		this.vertexStoreEventSender = Objects.requireNonNull(vertexStoreEventSender);
+		this.bftUpdateSender = Objects.requireNonNull(bftUpdateSender);
+		this.counters = Objects.requireNonNull(counters);
+		this.rootVertex = Objects.requireNonNull(rootVertex);
+		this.highestQC = Objects.requireNonNull(rootQC);
```
Comment (siy) [inline radixdlt/src/main/java/com/radixdlt/consensus/bft/VertexStore.java:81]: Checks above may throw exception, so they should be shifted to factory method and this constructor then will be just a plain all-args constructor with no chances to throw exception.

**Review by siy — APPROVED** [review]
LGTM

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java [hunk 2]
```diff
@@ -248,25 +248,35 @@ private void doCommittedSync(SyncState syncState) {
...
 		}
-		int nextIndex = random.nextInt(syncRequestState.authors.size());
-		BFTNode nextNode = syncRequestState.authors.get(nextIndex);
-		this.requestSender.sendGetVerticesRequest(nextNode, request);
-		this.timeoutScheduler.scheduleTimeout(request, bftSyncPatienceMillis);
+
+		List<HashCode> syncIds = syncRequestState.syncIds.stream().filter(syncing::containsKey).collect(Collectors.toList());
+		for (HashCode syncId : syncIds) {
+			SyncState syncState = syncing.get(syncId);
+			if (syncState == null) {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java]: Is it possible to get a null here, given the stream filter a couple of lines above?

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java [hunk 3]
```diff
@@ -248,25 +248,35 @@ private void doCommittedSync(SyncState syncState) {
...
+				return;
+			}
+
+			// Retry full sync on timeout
+			int nextIndex = random.nextInt(syncRequestState.authors.size());
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java:261]: I think we have a problem here.  I can't find a binding for Random.class, which means that this is going to be bound to a default constructed version of java.util.Random.

We should be using a binding to SecureRandom somewhere.

**Review by clstrfsck — CHANGES_REQUESTED** [review]
IMO we need to do something about binding for Random, and also the `rootQC.equals(rootQC)` in VertexStore.

File: radixdlt/src/main/java/com/radixdlt/consensus/bft/VertexStore.java [hunk 4]
```diff
@@ -66,36 +65,52 @@
...
+		if (!rootQC.getProposed().getVertexId().equals(rootVertex.getId())) {
+			throw new IllegalStateException(String.format("rootQC=%s does not match rootVertex=%s", rootQC, rootVertex));
+		}
 
-		Objects.requireNonNull(rootVertex);
-		Objects.requireNonNull(rootQC);
-		Objects.requireNonNull(vertices);
+		final Optional<BFTHeader> header = rootQC.getCommittedAndLedgerStateProof().map(Pair::getFirst);
+		if (!header.isPresent()) {
+			if (!rootQC.getView().isGenesis() || !rootQC.equals(rootQC)) {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/bft/VertexStore.java]: `rootQC.equals(rootQC)` looks either incorrect, or not required here.

Also, this clause and the next clause don't appear to be covered by unit tests.

**Review by LukasGasior1 — APPROVED** [review]

File: radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java (hunk 3 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/sync/BFTSync.java:261]: yup good point

File: radixdlt/src/main/java/com/radixdlt/SystemModule.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import com.radixdlt.network.TimeSupplier;
+import java.security.SecureRandom;
+import java.util.Random;
+import org.radix.time.Time;
+
+/**
+ * Module which specifies implementations of system objects such as
+ * random and time.
+ */
+public class SystemModule extends AbstractModule {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/SystemModule.java:33]: Moar gooder.

**Review by clstrfsck — APPROVED** [review]
Looks good.

**Pull request merged**